zvfs: move file ops from POSIX into ZVFS and move fdtable into ZVFS

fkokosinski opened the pull request:
This PR:
* moves all operations on single files into ZVFS and makes the POSIX subsystem call into ZVFS to perform them
* simplifies cooperation between different file access APIs (current & future)
* makes preparation for the file descriptor table becoming optional

It was necessary to define a `struct zvfs_stat` to avoid a dependency cycle. Functions used internally for file i/o operations are publicised since they won't require any changes between various subsystems. This allows ZVFS to actually fulfill its purpose of facilitating cooperation of different file APIs.

The file descriptor table is used in every area that expects to work on files through descriptor indices. It can only be operated on through functions whose names indicate a relationship with ZVFS (`zvfs_*fd*`). The integer file descriptor mechanism shouldn't be separate from ZVFS.

Commit:
zvfs: Move file operations from POSIX into ZVFS

This commit moves all operations on single files into ZVFS and makes the
POSIX subsystem call into ZVFS to perform them. It was necessary to define
a `struct zvfs_stat` to avoid a dependency cycle. Functions used
internally for file i/o operations are publicised since they won't require
any changes between various subsystems. This allows ZVFS to actually
fulfill its purpose of facilitating cooperation of different file APIs.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

Commit:
zvfs: Move fdtable into ZVFS

The file descriptor table is used in every area that expects to work on
files through descriptor indices. It can only be operated on through
functions whose names indicate a relationship with ZVFS (`zvfs_*fd*`).
The integer file descriptor mechanism shouldn't be separate from ZVFS.
This will make cooperation between different file access APIs much
simpler. This commit also makes preparations for the fdtable becoming
optional.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

jukkar commented on lib/os/fdtable.c:
```diff
@@ -23,8 +24,9 @@
+#include <zephyr/fs/fs.h>
 
-struct stat;
+K_MEM_SLAB_DEFINE(file_desc_slab, sizeof(struct fs_file_t), CONFIG_ZVFS_OPEN_MAX, 4);
```
Note that after the commit 1cfdf7c8d1c4b0a5e5a88d5b5c01ce7fb84152c9, you need to use `ZVFS_OPEN_SIZE` define instead of `CONFIG_ZVFS_OPEN_MAX` (as the latter is set to 0 in typical use cases).

Commit:
zvfs: Remove test for static initialization of stdin, stdout, stderr

This removes a function that created a new mutex and conditional variable
and used memcpy() to compare them with ones in a given fdtable entry.
Since those struct members are initialized statically, this test doesn't
serve much of a purpose anymore. Moreover, padding bytes inside structs
are technically not required to be zero, so these memcpy() calls caused
SonarQube to complain.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

de-nordic reviewed (commented):
[Once again](https://github.com/zephyrproject-rtos/zephyr/pull/96278#issuecomment-3361818767) why are we investing into this?
I do not see [any reason for ZVFS and POSIX to coexist in one system](https://github.com/zephyrproject-rtos/zephyr/pull/96278#issuecomment-3364918881).
Has anyone noticed that this is basically requiring something like 4-5 stack frames from POSIX to FS to even work? Who does it make POSIX work for? Who is the ZVFS for?

fkokosinski commented:
@de-nordic -

Thanks a lot for taking a look at this PR.

> [Once again](https://github.com/zephyrproject-rtos/zephyr/pull/96278#issuecomment-3361818767) why are we investing into this?

The POSIX subsystem has its users (including us) - so I suppose it's natural that those users will try to contribute patches/fixes they're using downstream that they found useful.

This PR is an attempt at addressing the comments we've received in the [previous PR](https://github.com/zephyrproject-rtos/zephyr/pull/96278). We'd be happy to address any further technical comments you might have. 

> I do not see any reason for ZVFS and POSIX to coexist in one system.

As you've noticed in the linked PR, ZVFS claims to be a [central](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ZVFS) library, and "is designed to be used by all Zephyr subsystems that need to work with files."

I'm not sure why the POSIX layer wouldn't fit into that description?

> Has anyone noticed that this is basically requiring something like 4-5 stack frames from POSIX to FS to even work? Who does it make POSIX work for? 

The POSIX layer is completely optional - it's up to the user whether they want to allocate the device's resources on it or not.

Plus, there are features in Zephyr that are not targeted towards the more resource-constrained devices, such as SMP. Zephyr is very useful and works well on beefy RISC-V or Armv8-A machines. There are also features that straight up wouldn't work on a physical device, like VIRTIO, but are useful nevertheless.

> Who is the ZVFS for?

At least according to the documentation, it's meant to be central, so for everyone.

Let me reiterate that we're committed to making these changes of as little impact to the average user as possible, meaning - if there's a change introduced by this PR that can be made optional, we'd be more than happy making it disabled by default.

de-nordic commented on lib/os/zvfs/zvfs_file_vmethods.c:
```diff
@@ -0,0 +1,144 @@
+ * @brief Read from a file.
+ *
+ * See IEEE 1003.1
+ */
```
Seems that ZVFS is already POSIX.

fkokosinski replied:
This is selectable by `ZVFS_DEFAULT_FILE_VMETHODS`, which is n-selected by default

de-nordic replied:
Still, it is already POSIX, does not matter how many Kconfigs are there needed to select this.
